### PR TITLE
Add functional tests #16

### LINKED ISSUE #16: Add functional tests
It should test that the `Backend` section displays correct information for the test used server, DB

### PR DESCRIPTION
**Pull request by ChiuchiuSorin**
Create functional tests for Admin Tools

### CODE REVIEW

**Comment by ChiuchiuSorin** [comment]
Currently having an issue with the following [test](https://github.com/xwikisas/application-admintools/pull/25/files#diff-9a24a56c8d4cfc69ff20d4fbab0164f8e55c3304f76af2117973119a5f56ded9R108-R116) at line 112. The issue is that the application is not seen in the Applications panel, but I'm not sure what causes it. I'm thinking it may be a missing dependency, but I did not identify what may be missing.

Error log: https://up1.xwikisas.com/#GN9TRHXpO49fjt0odyRZsw
Test screenshot: 
image

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsHomePage.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return "WebHome";
+    }
+
+    public static String getSpace()
+    {
+        return "AdminTools";
+    }
+
+    public static String getUrl()
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsHomePage.java:42]: We use upper case for acronyms.

```suggestion
    public static String getURL()
```

But I don't see where this method is used. In general, we shouldn't add APIs that are not used.

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsHomePage.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public static String getPage()
+    {
+        return "WebHome";
+    }
+
+    public static String getSpace()
+    {
+        return "AdminTools";
+    }
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsHomePage.java:40]: I wouldn't expose this. I'd rather have instead a method like ``isCurrentPage()``.

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+package com.xwiki.admintools.test.po;
+
+import java.util.List;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.xwiki.test.ui.po.ViewPage;
+
+public class AdminToolsViewPage extends ViewPage
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java:29]: Missing JavaDoc. Without it, it's not clear what's the difference between ``AdminToolsHomePage`` and ``AdminToolsViewPage``.

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+{
+    public List<WebElement> dashboardElements =
+        getDriver().findElements(By.xpath("//div[contains(@class, 'adminToolsDashboardItem')]"));
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java:32]: ``getDriver().findElements()`` is normally used only inside methods, to get the elements dynamically. For class fields we have ``@FindBy`` which works also on lists (check existing tests). Moreover, CSS selectors are simpler and easier to read in most cases. We normally rely on XPath selectors only for edge cases where a CSS selector can't be written.

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    @FindBy(xpath = "//a[@data-target = '#configurationViewLastNLinesModal']")
+    public WebElement backendLogsHyperlink;
+
+    @FindBy(xpath = "//div[@id = 'adminToolsBackendSection'] /ul")
+    public WebElement backendContent;
+
+    @FindBy(xpath = "//div[@id = 'adminToolsFilesSection'] /ul")
+    public WebElement filesContent;
+
+    public List<WebElement> getDashboardElements()
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java:46]: A page object shouldn't return a WebElement unless it's a simple element, like a link or an image. For complex elements (that represent a UI component like a dashboard gadget) it should return another page object that allows you to interact with that UI component. Don't we have page objects for the Dashboard application?

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java:1]: I feel this class does to much. It would be better to have dedicated page objects for all UI components from the admin tools page (e.g. for various admin tools dashboard gadgets).

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/DownloadArchiveModalView.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.admintools.test.po;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.xwiki.test.ui.po.ViewPage;
+
+public class DownloadArchiveModalView extends ViewPage
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/DownloadArchiveModalView.java:27]: Why extend ``ViewPage`` when we have ``BaseModal``? Doesn't this represent only the modal?

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/DownloadArchiveModalView.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+{
+    @FindBy(xpath = "//div[@id = 'downloadFilesModal']")
+    public WebElement content;
+
+    public String getModalText()
+    {
+        return content.getText();
+    }
+
+    public WebElement getXWikiConfigCheck()
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/DownloadArchiveModalView.java:37]:
```suggestion
    public WebElement getXWikiConfigCheckbox()
```

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/LastNLinesModalView.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.admintools.test.po;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.xwiki.test.ui.po.BaseElement;
+
+public class LastNLinesModalView extends BaseElement
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/LastNLinesModalView.java:27]: Again, why not extend ``BaseModal``?

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 9]
```diff
@@ -41,6 +41,7 @@ AdminTools.Code.ConfigurationClass_excludedLines=Excluded lines
 AdminTools.Code.ConfigurationClass_excludedLines.hint=Lines containing sensitive info in xwiki.cfg and xwiki.properties files
 AdminTools.Code.ConfigurationClass_serverLocation=Server location
 AdminTools.Code.ConfigurationClass_serverLocation.hint=Path to the server location
+AdminTools.home.title=Admin Tools
```
Comment (mflorea) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:44]:
```suggestion
adminTools.home.title=Admin Tools
```

The previous line uses upper case because it's a class name (reference).

File: application-admintools-ui/src/main/resources/AdminTools/WebHome.xml [hunk 10]
```diff
@@ -31,7 +31,7 @@
   <author>xwiki:XWiki.Admin</author>
   <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
   <version>1.1</version>
-  <title>Admin Tools</title>
+  <title>$services.localization.render("AdminTools.home.title")</title>
```
Comment (mflorea) [inline application-admintools-ui/src/main/resources/AdminTools/WebHome.xml:34]:
```suggestion
  <title>$services.localization.render("adminTools.home.title")</title>
```

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+    private static final String USER_NAME = "JonSnow";
+
+    private static final DocumentReference ADMINTOOLS_CONFIGURATION_REFERENCE =
+        new DocumentReference("xwiki", Arrays.asList("AdminTools", "Code"), "Configuration");
+
+    private static final String ADMINTOOLS_CONFIGURATION_CLASSNAME = "AdminTools.Code.ConfigurationClass";
+
+    private static final String BACKEND_SECTION_VIEW_LAST_LOGS_MODAL_ID = "#configurationViewLastNLinesModal";
+
+    private static final String DOWNLOAD_FILES_MODAL_ID = "#downloadFilesModal";
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:63]: This doesn't belong here in the test. It should be in a page object. Tests shouldn't use (directly) any selectors normally.

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+        setup.setGlobalRights("XWiki.XWikiAdminGroup", "", "admin", true);
+        setup.createAdminUser();
+        setup.loginAsAdmin();
+    }
+
+    @BeforeEach
+    void goToPage(TestUtils testUtils)
+    {
+        AdminToolsHomePage page = AdminToolsHomePage.gotoPage();
+        page.waitUntilPageIsReady();
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:105]: Isn't this already done by ``BasePage``?

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+        assertEquals(AdminToolsHomePage.getPage(), vp.getMetaDataValue("page"));
+    }
+
+    @Test
+    @Order(1)
+    void adminToolsHomePageBackend(TestConfiguration testConfiguration)
+    {
+        if (!isSupportedServer) {
+            return;
+        }
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:123]: This is not the right way. See https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-export/xwiki-platform-export-pdf/xwiki-platform-export-pdf-test/xwiki-platform-export-pdf-test-docker/src/test/it/org/xwiki/export/pdf/test/ui/PDFExportExecutionCondition.java . And you can also force a specific servlet container if you want with https://dev.xwiki.org/xwiki/bin/view/Community/Testing/DockerTesting/#HExample2:ChoosingContainer2BDB2BBrowserandversions .

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+        AdminToolsViewPage webHomePage = new AdminToolsViewPage();
+        WebElement viewLastLogsHyperlink = webHomePage.getBackendLogsHyperlink();
+        viewLastLogsHyperlink.click();
+        LastNLinesModalView lastLinesModalView = new LastNLinesModalView();
+        lastLinesModalView.waitUntilPageIsReady();
+        String mainWindowHandle = testUtils.getDriver().getWindowHandle();
+        lastLinesModalView.clickButton(BACKEND_SECTION_VIEW_LAST_LOGS_MODAL_ID);
+        switchToNewTab(testUtils, mainWindowHandle);
+        testUtils.getDriver().switchTo().window(mainWindowHandle);
```
Comment (mflorea) [inline application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:155]: Why do you need to switch tabs?

**Review by mflorea — CHANGES_REQUESTED** [review]

File: application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java (hunk 3 above)
Comment (ChiuchiuSorin) [reply application-admintools-test/application-admintools-test-pageobjects/src/main/java/com/xwiki/admintools/test/po/AdminToolsViewPage.java:29]: Following your suggestion below, I removed the `AdminToolsViewPage` and split it in classes that handle the different sections of the dashboard.

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java (hunk 14 above)
Comment (ChiuchiuSorin) [reply application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:155]: I need to manually switch tabs to verify the opened content, as Selenium does not change focus when a tab is opened. I had a [comment](https://github.com/xwikisas/application-admintools/pull/25/files#diff-9a24a56c8d4cfc69ff20d4fbab0164f8e55c3304f76af2117973119a5f56ded9R260-R262) explaining this but did not offer a lot of details.

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java (hunk 13 above)
Comment (ChiuchiuSorin) [reply application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:123]: I don't want to force a specific container, as in the future multiple servlet engines will be supported. But I added the condition to check if the used container is currently supported.

File: application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
+    {
+        ApplicationsPanel applicationPanel = ApplicationsPanel.gotoPage();
+        ViewPage vp = applicationPanel.clickApplication("Admin Tools");
+        Assertions.assertTrue(AdminToolsHomePage.isCurrentPage(vp));
+    }
+
+    @Test
+    @Order(2)
+    void backendSectionTest(TestConfiguration testConfiguration)
+    {
```
Comment (sarahxb) [inline application-admintools-test/application-admintools-test-docker/src/test/it/com/xwiki/admintools/test/ui/AdminToolsIT.java:131]: I would maybe leave a comment here explaining what's tested exactly (and in the following tests as well )

**Review by sarahxb — CHANGES_REQUESTED** [review]
Overall the functional tests part looks good to me, I just feel that some more comments for the less intuitive tests would be useful.

**Review by sarahxb — APPROVED** [review]

**Review by oanalavinia — APPROVED** [review]
